core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapCastInComparison.java

```diff
@@ -159,36 +163,153 @@ public Visitor(PlannerContext plannerContext, Session session)
         public Expression rewriteComparison(Comparison node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
         {
             Comparison expression = treeRewriter.defaultRewrite(node, null);
-            return unwrapCast(expression);
+            return unwrapCast(expression.operator(), expression.left(), expression.right()).orElse(expression);
         }
 
-        private Expression unwrapCast(Comparison expression)
+        @Override
+        public Expression rewriteBetween(Between node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
         {
-            // Canonicalization is handled by CanonicalizeExpressionRewriter
-            if (!(expression.left() instanceof Cast cast)) {
+            Between expression = treeRewriter.defaultRewrite(node, null);
+
+            if (!(expression.value() instanceof Cast cast)) {
+                return expression;
+            }
+
+            Optional<Expression> optionalLowBoundUnwrapped = unwrapCast(GREATER_THAN_OR_EQUAL, node.value(), node.min());
+            Optional<Expression> optionalHighBoundUnwrapped = unwrapCast(LESS_THAN_OR_EQUAL, node.value(), node.max());
+            if (optionalLowBoundUnwrapped.isEmpty() || optionalHighBoundUnwrapped.isEmpty()) {
                 return expression;
             }
+            Expression lowBoundUnwrapped = optionalLowBoundUnwrapped.get();
+            Expression highBoundUnwrapped = optionalHighBoundUnwrapped.get();
+
+            Expression trueIfNotNullExpression = trueIfNotNull(cast.expression());
+            if (trueIfNotNullExpression.equals(lowBoundUnwrapped)) {
+                return highBoundUnwrapped;
+            }
+            if (trueIfNotNullExpression.equals(highBoundUnwrapped)) {
+                return lowBoundUnwrapped;
+            }
```

Comment:
Not that this is potentially incorrect. Consider
- `lowBoundUnwrapped`: `trueIfNotNull(cast)`
- `highBoundUnwrapped`: evaluates to `true` for null value of `cast`.

I understand this is not the case today because the `lowBoundUnwrapped` and `highBoundUnwrapped` can only take a few different forms, as they are results of the `unwrapCast` method.

However, the `unwrapCast` method might evolve over time, and this logic might get out of sync.

I suggest that before we proceed to combining `lowBoundUnwrapped` and `highBoundUnwrapped`, we first assert that both expressions satisfy some structural requirements.